go-controller/pkg/node/gateway_shared_intf.go

```diff
@@ -154,76 +285,268 @@ func (npw *nodePortWatcher) updateServiceFlowCache(service *kapi.Service, add bo
+	out, exists = npw.serviceInfo[index]
+	delete(npw.serviceInfo, index)
+	return out, exists
+}
+
+// getServiceInfo returns the serviceConfig for a service and if it exists
+func (npw *nodePortWatcher) getServiceInfo(index ktypes.NamespacedName) (out *serviceConfig, exists bool) {
+	npw.serviceInfoLock.Lock()
+	defer npw.serviceInfoLock.Unlock()
+	out, exists = npw.serviceInfo[index]
+	return out, exists
+}
+
+// setServiceInfo creates and sets the serviceConfig
+func (npw *nodePortWatcher) setServiceInfo(index ktypes.NamespacedName, service *kapi.Service, etpHostRules bool) {
+	npw.serviceInfoLock.Lock()
+	defer npw.serviceInfoLock.Unlock()
+
+	npw.serviceInfo[index] = &serviceConfig{service: service, etpHostRules: etpHostRules}
+}
+
+// addOrSetServiceInfo creates sets the serviceConfig, if it already exists do nothing
+func (npw *nodePortWatcher) addOrSetServiceInfo(index ktypes.NamespacedName, service *kapi.Service, etpHostRules bool) (exists bool) {
+	npw.serviceInfoLock.Lock()
+	defer npw.serviceInfoLock.Unlock()
+
+	if _, exists := npw.serviceInfo[index]; !exists {
+		// Only Set this if it dosn't exist
+		npw.serviceInfo[index] = &serviceConfig{service: service, etpHostRules: etpHostRules}
+		return false
+	}
+	return true
+
+}
+
+// updateServiceInfo sets the serviceConfig for a service and returns the existing serviceConfig, if inputs are nil
+// do not update those fields, if it does not exist return nil.
+func (npw *nodePortWatcher) updateServiceInfo(index ktypes.NamespacedName, service *kapi.Service, etpHostRules *bool) (old *serviceConfig) {
+	var exists bool
+
+	npw.serviceInfoLock.Lock()
+	defer npw.serviceInfoLock.Unlock()
+
+	if old, exists = npw.serviceInfo[index]; exists {
+		klog.V(5).Infof("No serviceConfig found for service %s in namespace %s", index.Name, index.Namespace)
+		return nil
+	}
+
+	old = npw.serviceInfo[index]
+
+	if service != nil {
+		npw.serviceInfo[index].service = service
+	}
+
+	if etpHostRules != nil {
+		npw.serviceInfo[index].etpHostRules = *etpHostRules
+	}
+
+	return old
+}
+
+// addServiceRules ensures the correct iptables rules and OpenFlow physical
+// flows are programmed for a given configuration
+func (npw *nodePortWatcher) addServiceRules(service *kapi.Service, hasHostNet bool) {
+	if hasHostNet && util.ServiceExternalTrafficPolicyLocal(service) {
+		npw.updateServiceFlowCache(service, true, hasHostNet)
+		npw.ofm.requestFlowSync()
+		addSharedGatewayIptRules(service, hasHostNet)
+	} else {
+		npw.updateServiceFlowCache(service, true, hasHostNet)
+		npw.ofm.requestFlowSync()
+		addSharedGatewayIptRules(service, hasHostNet)
+	}
+}
+
+// delServiceRules ensures the correct iptables rules and OpenFlow physical
+// flows are deleted for a given previous configuration
+func (npw *nodePortWatcher) delServiceRules(service *kapi.Service, hasHostNet bool) {
+	if hasHostNet && util.ServiceExternalTrafficPolicyLocal(service) {
+		npw.updateServiceFlowCache(service, false, hasHostNet)
```

Comment:
updateServiceFlowCache deletes by cookie right? So it shouldnt matter if it had host net or not when you are deleting?